lib/AST/GenericSignature.cpp

```diff
@@ -938,7 +946,12 @@ void GenericSignature::verify(ArrayRef<Requirement> reqts) const {
           llvm::errs() << "\n";
           abort();
         }
-        if (compareDependentTypes(firstType, secondType) >= 0) {
+        // FIXME: Same-element rewrite rules are of the form [element].T => U,
+        // which turns into a same-type requirement repeat U == each T. The
+        // pack element must always be on the left side of the rule, so type
+        // parameters can appear out of order in the final same-type requirement.
+        if (firstType->isParameterPack() == secondType->isParameterPack() &&
```

Comment:
Don't forget to remove this part -- we can't change type parameter order because it will break ABI

Reply:
Sima and I were discussing this and we're a bit confused. This code is just in `GenericSignature::verify`; I don't think this changes type parameter order of existing code. My original implementation did change the order initially but I had fixed this in https://github.com/apple/swift/pull/67465/commits/d138b1553f776f4ae79cea8484c3129fc7d3c4a9.

For a same-element requirement, my understanding is we need the element parameter to be on the left because the reduced type of a pack with a same-element requirement has to be the scalar type, which is why we have this special handling for the order. I see in my original PR you left a comment that the requirement builder should reorder them. Is your suggestion that the rewrite system should order the element terms before pack terms, but we should change that when covering it to a generic signature?

Reply:
Okay Slava helped me remember what's going on here: We can fix the `getReducedType` query to always return the scalar type as the reduced type with a dedicated `getReducedElementType` query that's only ever called on the pattern of a pack expansion. We don't necessarily want `getReducedType` to have this behavior for packs in general, because shape types are also represented with parameter packs, which should not be reduced to a scalar type. So we don't necessarily need the minimized generic signature to have this order; we just need the requirement machine and `getReducedElementType` query to handle that. The reduced type query already handles pack expansion types specially to call the `getReducedShape` query, so that's where we'd need to call `getReducedElementType` too.

Reply:
It would be simplest to maintain the invariant that for a same-type requirement `T == U` we have T < U, so that means a same-element requirement is either going to have the pack on the left or the right depending on the order of the two type parameters. In the reduction order though, the term `[element].T` is always greater than `U` for any `U` not involving `[element]`, so when forming requirements from rules we might need to swap the two sides, depending on which one is longer once we drop the `[element]`, essentially.